### DISCUSSION TITLE
micropython/lora sync word

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13309] [Libraries & Drivers] [answered]
The README.md says that it is set to a default of 0x12.  Where does this happen?

**Comment by davefes** [comment]
I guess all of this:
` syncword = 0x0404 + ((syncword & 0x0F) << 4) + ((syncword & 0xF0) << 8)
gives you 0x12

**Reply by robert-hh** [reply]
> `syncword = 0x0404 + ((syncword & 0x0F) << 4) + ((syncword & 0xF0) << 8)` gives you 0x12

Hardly with this expression. Since values are added to 0x0404, the result cannot be less than 0x0404.

**Comment by davefes** [comment]
Good call.  I'll re-open because I have searched through all the source code and can not find where the default is set.

**Comment by sophiedegran** [comment]
Dave
Can't find anywhere too, it seem to be in lora_cfg ...
But as sx126x boards used a different sync word look at sx126x.py and they force the board to change from 0x12 to 0x1424 as it should be (private lora network).
Hope that help ☺

**Reply by davefes** [reply]
` if "syncword" in lora_cfg:` in sx126x.py only gets called if you put `syncword ` into lora_cfg.  It does seem to be somewhere as two units will actually talk to each other without putting 0x12 into lora_cfg.

**Reply by sophiedegran** [reply, marked as answer]
Yes you're right but I think lora_cfg is created at the init of the lora modul and there is a 'get_configuration' (as in others lora library) which implemented it.. As lora boards are all differents, the 0x12 is inside the lora modul (mostly semtech ..) and for sx126x boards it's 0x1424 (16 bits of 0x12) 
For lorawan, lora commercial network it's 0x34 (or 0x3444 16bits)

**Reply by davefes** [reply]
Where is `get_configuration`?  Have you tried micropython/lora?  I can get `my_irq_callback()` to work using `modem.poll_recv()` but can not stop the receiver so that I can send a message.

Was straight-forward using Wei's implementation:
https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master

**Reply by sophiedegran** [reply]
get_configuration is for Ebyte lora library micropython that I use ... You can check this library which work perfectly.. 
As I have only Ebyte boards because I use them on long distance (5-10km), I use only this which work perfectly well

**Reply by davefes** [reply]
OK, thanks for that.  I'll take "it is in the RFM96W, which probably uses a Semtech chip" as an answer.

**Reply by projectgus** [reply]
@sophiedegran has answered this 100% correctly already, but just to spell out the fiddly nuances:

* The sync words have to match for radios to be able to communicate with each other.
* SX126x and SX127x chip families store the same LoRa Sync Word using a different hardware register representation (unclear why).
* The default hardware sync word register values (from reset) are 0x1424 for SX126x and 0x12 for SX127x. These represent the same LoRa sync word.
* This is why in MicroPython the radio will use this sync word value even if you don't add the [syncword key in the `lora_cfg` dict](https://github.com/projectgus/micropython-lib/blob/f816c968e081181f4ead241ab55d589bcda22bee/micropython/lora/README.md#syncword---sync-word), the driver leaves it at the default value in hardware.
* Sync word 0x34 (aka 0x3444) is reserved for LoRaWAN networks. The SX126x datasheet simply says "Set to 0x3444 for Public Network. Set to 0x1424 for Private Network". I think it's technically possible to set other values (some bit patterns may not work as well as others), although not recommended.

**Reply by davefes** [reply]
Thank you for expanding on that.  Looking through all the RFM9W data the ;one of a few `0x12` is the version number of the chip.  Seems it could in co-incidental.

**Comment by projectgus** [comment]
I must apologise, at the time this question was asked `syncword` functionality did not work as documented in any of the LoRa drivers! You would get the "public" sync word no matter what value was set in the configuration.

The fixes for this are included here:
https://github.com/micropython/micropython-lib/pull/804
